### PR TITLE
 #84 Feat 84 RDS Creation Terraform

### PR DESCRIPTION
**Pull request by lit-dds**
Thanks for submitting a pull request! Below are a few things you can do to help us more quickly review your changes.

#84 

@dds-hansolo peered program along the way.

## Checklist

I have…

- [ x ] summarized below my changes and noted which issues (if any) this pull request fixes or addresses.
- [ x ] thoroughly outlined below the steps necessary to test my changes.
- ~~[ ] attached screenshots illustrating relevant behavior before and after my changes.~~
- [ x ] review and update SSAG documentation if needed.

## Summary of Changes

This pull request…

- Refractor the Terraform Infrastructure, Platform, and Application folder to accommodate the Relational Database Service dual availability zone requirements for High Availability and High Scalability.
- Pulled database configurations from the Platform folder to its own separate Database folder to solely create or modify the database. This is to "loosely couple" the architecture from the database so making changes to the overall architecture does not destroy the database each time.
- Delete variables.tf, add variables to top of main terraform files, and centrally located editable values to .tfvars file.
- Reconfigure deploy.sh for running in circleci from scanning, building, pushing, and creating/modifying infrastructure.
- update Elastic Container Service task definition to take a sha256 digest as the container tag to verify if the service needs to rerun or not.

## Testing

To verify the changes proposed in this pull request…

1. `terraform fmt` to rewrite files to a canonical format and style which applies the Terraform language style conventions along with minor adjustments for readability.
1. `terraform validate` command validates the files in each directory, referring only to the config and not accessing any remote services such as remote state in AWS.
1. `tflint` command is a linter that focuses on possible errors and best practices.
1. Access AWS console and verify configurations on the graphical user interface.
1. Access defined dns in an incognito browser.

## Screenshots

N/A

### CODE REVIEW

File: deploy/staging/database/database.tf [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  max_allocated_storage = 1000
+  storage_type          = "gp2"
+  engine                = "postgres"
+  engine_version        = "11.5"
+  //  storage_encrypted      = true
+  skip_final_snapshot       = true
+  final_snapshot_identifier = "solo-stage-tf-db-micro"
+
+  //  DO NOT COMMIT THIS
+  name     = "solodb"
```
Comment (jhackshaw-dds) [inline deploy/staging/database/database.tf]: forgot to delete comment?

File: deploy/staging/application/application-stage.tfvars [hunk 2]
```diff
@@ -1,5 +1,16 @@
-// Remote State
-remote_state_bucket = "solo-stage-ecs-fargate-terraform-remote-state"
-remote_state_key    = "STAGE/platform-stage.tfstate"
-
-internet_cidr_block = "0.0.0.0/0"
\ No newline at end of file
+remote_state_bucket       = "solo-stage-ecs-fargate-terraform-remote-state"
+remote_state_key          = "STAGE/platform-stage.tfstate"
+project                   = "SOLO"
+internet_cidr_block       = "0.0.0.0/0"
```
Comment (jhackshaw-dds) [inline deploy/staging/application/application-stage.tfvars]: is this variable used? can it just be hardcoded?

File: deploy/staging/application/application-stage.tfvars [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+region                    = "us-gov-west-1"
+solo_iam_role_name        = "solo-fargate"
+ecs_task_def_service_name = "solo_stage_tf_service_application"
+task_definition_name      = "solo_stage_tf_task"
+solo_iam_task_exe_role    = "soloStageTaskExeRole"
+backend_container_name    = "solo-stage-tf-backend-container-name"
+frontend_container_name   = "solo-stage-tf-frontend-container-name"
```
Comment (jhackshaw-dds) [inline deploy/staging/application/application-stage.tfvars]: can we take the "-name" out of the name?

File: deploy/staging/application/app.tf [hunk 4]
```diff
@@ -16,73 +68,103 @@ data "terraform_remote_state" "platform_stage" {
...
   memory                   = var.memory
-  task_role_arn            = data.aws_iam_role.ecs_task_execution_role.arn
-  execution_role_arn       = data.aws_iam_role.ecs_task_execution_role.arn
+  task_role_arn            = data.aws_iam_role.ecs_task_role.arn
+  execution_role_arn       = data.aws_iam_role.ecs_task_exe_role.arn
 
   tags = {
-    Name    = "solo_stage_tf_task"
-    Project = "SOLO"
+    Name    = "solo_stage_tf_td_application"
```
Comment (jhackshaw-dds) [inline deploy/staging/application/app.tf:120]: I think there's a variable for this defined on line 15, it seems fine just hardcoded like this but then do we need the variable?

File: deploy/staging/database/database.tf [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  allocated_storage     = 20
+  max_allocated_storage = 1000
+  storage_type          = "gp2"
+  engine                = "postgres"
+  engine_version        = "11.5"
+  //  storage_encrypted      = true
+  skip_final_snapshot       = true
+  final_snapshot_identifier = "solo-stage-tf-db-micro"
+
+  //  DO NOT COMMIT THIS
```
Comment (jhackshaw-dds) [inline deploy/staging/database/database.tf]: Forgot to remove comment?

File: deploy/staging/application/containers_definition.json [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      {
+        "hostPort": 443,
+        "protocol": "tcp",
+        "containerPort": 443
+      },
+      {
+        "hostPort": 80,
+        "protocol": "tcp",
+        "containerPort": 80
+      }
```
Comment (jhackshaw-dds) [inline deploy/staging/application/containers_definition.json]: Is there ever a chance of port 80 making it to our frontend container for this to apply?

File: deploy/staging/database/database-stage.tfvars [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// Remote State
+remote_state_bucket = "solo-stage-ecs-fargate-terraform-remote-state"
+remote_state_key    = "STAGE/platform-stage.tfstate"
+region              = "us-gov-west-1"
+project             = "SOLO"
+
+internet_cidr_block = "0.0.0.0/0"
```
Comment (jhackshaw-dds) [inline deploy/staging/database/database-stage.tfvars]: recommend hardcoding this

File: gitignore [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+.python-version
+
+# binary file created by celery beat
+celerybeat-schedule
+
+.mypy_cache/
+
+# Terraform
+.terraform
+
```
Comment (jhackshaw-dds) [inline gitignore]: This file should be deleted

File: deploy/staging/database/database.tf (hunk 1 above)
Comment (lit-dds) [reply deploy/staging/database/database.tf]: comment delete

File: deploy/staging/application/application-stage.tfvars (hunk 3 above)
Comment (lit-dds) [reply deploy/staging/application/application-stage.tfvars]: yes. good call

File: deploy/staging/application/containers_definition.json (hunk 6 above)
Comment (lit-dds) [reply deploy/staging/application/containers_definition.json]: deleted. we don't need that object. it goes straight to 443

File: deploy/staging/database/database-stage.tfvars (hunk 7 above)
Comment (lit-dds) [reply deploy/staging/database/database-stage.tfvars]: yes to "region" and "cidr". no to "project". It gets used too often.

File: gitignore (hunk 8 above)
Comment (lit-dds) [reply gitignore]: don't know why this was ever created and pushed up...It is deleted

File: deploy/staging/README.MD [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+## Sequence of events
+The architecture is staged in `3.5` steps. The `3` main steps are infrastructure, 
+platform, and application. The `.5` is the database. Follow the instructions in 
+the README.MD in this folder order:
+1. Infrastructure
+1. Platform
+1. Application
+
+* Database is only used if there is a big change to the resources used for it. Also,
```
Comment (vcastro-DDS) [inline deploy/staging/README.MD:16]: Where do we have the logic that dictates what a big change to the resources is?

**Review by vcastro-DDS — APPROVED** [review]
Approve the PR. We peer coded several of the components. Lit successfully implemented architecture changes to use one availability zone (with another subnet in a different availability zone for Amazon RDS purposes). I left a comment about defining what database changes represent, and after further discussion, it is an automated process dependent merging to master and does not necessitate manual intervention. A change in state within the database config file would cause the database update to run via Terraform.

**Pull request merged**